Write-Through Inter-Block Cache

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

* Restructure `baseapp/` directory by logical grouping to improve readability.
* Improve `BaseApp` docs w/ diagrams 🎉 
* Implement `CommitKVStoreCacheManager` which acts as a persistent inter-block write-through cache.
  * It contains a mapping from `StoreKey` to a `*CommitKVStoreCache` (LRU [ARC](https://en.wikipedia.org/wiki/Adaptive_replacement_cache) cache).
  * A single instance of `CommitKVStoreCacheManager` is created in an app and set on the `BaseApp`.
  * The `BaseApp` sets this persistent cache on the root `CommitMultiStore`.
  * When the version is loaded, each IAVL `Store` is wrapped with it's own respective `*CommitKVStoreCache`.
  * Writes happen in a write-through fashion.

Preliminary benchmarks show there is only a marginal improvement.

closes: #1947

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

alexanderbez commented:
Update: The inter-block cache seems to work in practice (simulation and unit tests run successfully). However, after implementing some benchmarks, it seem while IO and memory improves noticeably, the CPU time is drastically worse. 

Looking at CPU profiling, it boils down to `Commit()` and I'm baffled because I don't see how/why this would be impacted...currently investigating.

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+
+// SetOption implements the ABCI interface.
+func (app *BaseApp) SetOption(req abci.RequestSetOption) (res abci.ResponseSetOption) {
+	// TODO: Implement!
```
issue ?

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+// EndBlock implements the ABCI interface.
+func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBlock) {
+	if app.deliverState.ms.TracingEnabled() {
+		app.deliverState.ms = app.deliverState.ms.SetTracingContext(nil).(sdk.CacheMultiStore)
```
why do we set the tracing context here to nil and on `BeginBlock` we provide a map? a comment on both cases would be useful

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+			}
+
+		default:
+			result = sdk.ErrUnknownRequest(fmt.Sprintf("Unknown query: %s", path)).Result()
```
```suggestion
			result = sdk.ErrUnknownRequest(fmt.Sprintf("unknown query: %s", path)).Result()
```

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+		}
+	}
+
+	msg := "Expected second parameter to be either simulate or version, neither was present"
```
```suggestion
	msg := "expected second parameter to be either 'simulate' or 'version', neither was present"
```

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+
+func handleQueryP2P(app *BaseApp, path []string, _ abci.RequestQuery) (res abci.ResponseQuery) {
+	// "/p2p" prefix for p2p queries
+	if len(path) >= 4 {
```
this should be the inverse:

```go
if len(path) < 4 {
        msg := "path should be p2p filter <addr|id> <parameter>"
	return sdk.ErrUnknownRequest(msg).QueryResult()
}

...

```

fedekunze replied:
#4965 4965

fedekunze commented on baseapp/abci.go:
```diff
@@ -0,0 +1,419 @@
+			}
+
+		default:
+			msg := "Expected second parameter to be filter"
```
```suggestion
			msg := "expected second parameter to be 'filter'"
```

fedekunze commented on baseapp/baseapp.go:
```diff
@@ -1030,67 +620,79 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-	// Commit. Use the header from this latest block.
-	app.setCheckState(header)
+	// NOTE: GasWanted is determined by ante handler and GasUsed by the GasMeter.
+	for msgIdx, msg := range msgs {
```
```suggestion
	for i, msg := range msgs {
```

fedekunze commented on baseapp/baseapp.go:
```diff
@@ -1030,67 +620,79 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
+		msgRoute := msg.Route()
+		handler := app.router.Route(msgRoute)
+		if handler == nil {
+			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgRoute).Result()
```
```suggestion
			return sdk.ErrUnknownRequest("unrecognized Msg type: " + msgRoute).Result()
```

fedekunze commented on baseapp/baseapp.go:
```diff
@@ -1030,67 +620,79 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-	ms  sdk.CacheMultiStore
-	ctx sdk.Context
-}
+		idxLog := sdk.ABCIMessageLog{MsgIndex: uint16(msgIdx), Log: msgResult.Log}
```
```suggestion
		idxLog := sdk.ABCIMessageLog{MsgIndex: uint16(i), Log: msgResult.Log}
```

fedekunze commented on baseapp/baseapp.go:
```diff
@@ -1030,67 +620,79 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-	return st.ctx
+// Splits a string path using the delimiter '/'.
+// e.g. "this/is/funny" becomes []string{"this", "is", "funny"}
+func splitPath(requestPath string) (path []string) {
```
this should be in utils

cwgoes commented on store/store.go:
```diff
@@ -11,6 +12,10 @@ func NewCommitMultiStore(db dbm.DB) types.CommitMultiStore {
 }
 
+func NewCommitKVStoreCacheManager() types.MultiStorePersistentCache {
+	return cache.NewCommitKVStoreCacheManager(cache.DefaultCommitKVStoreCacheSize)
```
Should this be a end-user-configurable paramater (in some `.toml` file)?

alexanderbez replied:
Added `--inter-block-cache` flag to the server package with `true` by default.

fedekunze commented on server/mock/store.go:
```diff
@@ -95,6 +95,10 @@ func (ms multiStore) GetStoreType() sdk.StoreType {
 	panic("not implemented")
 }
 
+func (ms multiStore) SetInterBlockCache(_ sdk.MultiStorePersistentCache) {
```
missing godoc comment: `SetInterBlockCache is not yet implemented`

fedekunze commented on store/cache/cache.go:
```diff
@@ -0,0 +1,112 @@
+	}
+)
+
+func NewCommitKVStoreCache(store types.CommitKVStore, size uint) *CommitKVStoreCache {
```
godoc comments missing in this file

fedekunze commented on store/cache/cache_test.go:
```diff
@@ -0,0 +1,60 @@
+		require.Equal(t, res, value)
+	}
+
+	for i := uint(0); i < cache.DefaultCommitKVStoreCacheSize*2; i++ {
```
can we move the contents into the prev loop?

fedekunze commented on store/store.go:
```diff
@@ -11,6 +12,10 @@ func NewCommitMultiStore(db dbm.DB) types.CommitMultiStore {
 	return rootmulti.NewStore(db)
 }
 
+func NewCommitKVStoreCacheManager() types.MultiStorePersistentCache {
```
godoc

fedekunze commented on x/mock/app.go:
```diff
@@ -206,9 +206,8 @@ func SetGenesis(app *App, accs []authexported.Account) {
-// GenTx generates a signed mock transaction.
-func GenTx(msgs []sdk.Msg, accnums []uint64, seq []uint64, priv ...crypto.PrivKey) auth.StdTx {
-	// Make the transaction free
+func GenTxWithChainID(cid string, msgs []sdk.Msg, accnums []uint64, seq []uint64, priv ...crypto.PrivKey) auth.StdTx {
```
these structs will be deleted for the ones that are located in `simapp/test_helpers.go`. We may want to dup this func there as well

cwgoes commented:
> > I'm surprised that benchmarks only show marginal improvements. What is the bottleneck now - Amino?
> 
> Perhaps I'm not creating the best benchmarks to illustrate the changes. I've tried a simple benchmark where I just commit `MsgSend` txs. I've also used the existing simulation benchmark. Do you have thoughts or suggestions? I can post a pprof if you'd like.

Actually what I would like to see are benchmarks of syncing a chain (e.g. the Cosmos Hub), if possible.

AdityaSripal commented on docs/core/baseapp.md:
```diff
@@ -7,167 +7,206 @@
-- `deliverState`: This state is updated during [`DeliverTx`](#delivertx), and reset on [`Commit`](#commit).
+- `checkState`: This state is updated during [`CheckTx`](#checktx-1), and reset on [`Commit`](#commit-1).
+- `deliverState`: This state is updated during [`DeliverTx`](#delivertx-1), and set to `nil` on
+[`Commit`](#commit-1).
```
```suggestion
[`Commit`](#commit-1). `deliverState` gets re-initialized on BeginBlock.
```

AdityaSripal commented on docs/core/baseapp.md:
```diff
@@ -7,167 +7,206 @@
+Note, the `deliverState` is set to `nil` on [`Commit`](#commit-1).
 
-Volatile - or cached - states are used in between [`Commit`s](#commit) to manage temporary states. They are reset to the latest version of the main state after it is committed. There are two main volatile states:
+![CheckTx](./baseapp_state-begin_block.png)
```
```suggestion
![BeginBlock](./baseapp_state-begin_block.png)
```

AdityaSripal commented on docs/core/baseapp.md:
```diff
@@ -7,167 +7,206 @@
+The state flow for `DeliverTx` is nearly identical to `CheckTx` except state transitions occur on
+the `deliverState` and messages in a transaction are executed. Similarly to `CheckTx`, state transitions
+occur on a doubly cache-wrapped state -- `deliverState`. Successful message execution results in
+writes being committed to `deliverState`.
```
Perhaps worth noting that even if msg execution fails here, any state updates from successful antehandler get persisted.

alexanderbez replied:
Updated.

AdityaSripal commented on store/cache/cache.go:
```diff
@@ -0,0 +1,112 @@
+	// in an inter-block (persistent) manner and typically provided by a
+	// CommitMultiStore.
+	CommitKVStoreCacheManager struct {
+		cacheSize uint
```
Would it be useful to be able to define cacheSize on a per-store basis?

For example, may not want a cache at all for accountStore. May want a bigger cache for govStore. A small but performant cache for params, etc.

alexanderbez replied:
Yeah so in theory that may prove to be valuable, but I hesitate to add such functionality in this initial PR for the following reasons:

- Complicates the API (although frivolous) 
- Requires fine benchmarking to know the values for each desired respective store

I recommend we create these APIs in a subsequent PR in a non-breaking fashion.

AdityaSripal commented on store/cache/cache.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// Reset resets in the internal caches.
+func (cmgr *CommitKVStoreCacheManager) Reset() {
```
When would this get used?

alexanderbez replied:
On `InitChain` (e.g. `TestAppSimulationAfterImport`).

AdityaSripal commented on store/cache/cache.go:
```diff
@@ -0,0 +1,112 @@
+
+	// cache miss; write to cache
+	value := ckv.CommitKVStore.Get(key)
+	ckv.cache.Add(string(key), value)
```
Should we be checking if `value != nil` first before adding?

alexanderbez replied:
The underlying `CommitKVStore` is guaranteed never to have a `nil` value. But i've added checks just to be safe.

tnachen commented on store/cache/cache.go:
```diff
@@ -0,0 +1,112 @@
+// If the value doesn't exist in the write-through cache, the query is delegated
+// to the underlying CommitKVStore.
+func (ckv *CommitKVStoreCache) Get(key []byte) []byte {
+	valueI, ok := ckv.cache.Get(string(key))
```
I don't know how long key is, but should consider storing string(key) as a variable as you can avoid the second string(key) in Add, so saves you going through the byte array twice.

alexanderbez replied:
String type casting is extremely negligible as far as overall CPU time is concerned, but I've added it to a variable as it doesn't hurt readability.

tnachen commented on store/iavl/store.go:
```diff
@@ -22,8 +22,34 @@ const (
+	// A value of 0 means store no waypoints. (node cannot assist in state-sync)
+	// By default this value should be set the same across all nodes,
+	// so that nodes can know the waypoints their peers store.
+	storeEvery int64
```
Why not use a boolean here if we're only having two possible values?

alexanderbez replied:
Not sure I follow? This was not introduced in this PR and relates to pruning. See [here](https://github.com/cosmos/cosmos-sdk/blob/v0.37.0/store/iavl/store.go#L111-L121) for pruning usage.

AdityaSripal replied:
I don't think this is supposed to be only two values. If the snapshotting frequency is every 10,000 blocks for state-sync, then a node can specify that they want to store every 10,000th block.

This will minimize storage requirements while still allowing node to participate in state-sync

tnachen replied:
Got it, sorry I didn't realize this was just moved from somewhere else.

tnachen commented on x/simulation/account.go:
```diff
@@ -38,13 +37,8 @@ func RandomAccounts(r *rand.Rand, n int) []Account {
 		// don't need that much entropy for simulation
 		privkeySeed := make([]byte, 15)
 		r.Read(privkeySeed)
-		useSecp := r.Int63()%2 == 0
```
Is this related to the cahce change?

alexanderbez replied:
Not directly, but was needed for initial benchmarking which I henceforth removed. So technically I can undo this change, but we only support secp256k1 keys anyway. Might as well keep it.

alexanderbez commented:
@cwgoes I've shared benchmarks and we can see this does indeed offer great improvement, specifically within `BeginBlock`.

I've addressed all comments @AdityaSripal @tnachen 👍
